core/arch/interp.c

```diff
@@ -817,6 +817,27 @@ check_new_page_jmp(dcontext_t *dcontext, build_bb_t *bb, app_pc new_pc)
 }
 
 static inline void
+bb_process_single_step(dcontext_t *dcontext, build_bb_t *bb)
+{
+    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n", bb->instr_start);
+    instrlist_append(bb->ilist, bb->instr);
+    /* Sets exit reason dynamically. */
+    instrlist_preinsert(bb->ilist, bb->instr,
```

Comment:
About `dcontext->upcontext.upcontext.exit_reason`, I did not know it should be set dynamically. I changed it this way.